Feature/add stakeholder modal

samshara opened the pull request:
fixes #1518

samshara committed:
Create AddStakeholderModal to add stakeholders to a project

samshara committed:
Add StakeholderList component

samshara committed:
Use AddStakeholderModal in ProjectDetailsForm component

AdityaKhatri commented on src/components/general/AddStakeholderButton/index.tsx:
```diff
@@ -0,0 +1,54 @@
+        setModalHidden,
+    ] = useModalState(false);
+
+    const handleShowStakeholdersModal = useCallback(() => {
+        if (showModal) {
+            setModalHidden();
+        } else {
+            setModalVisible();
+        }
+    }, [setModalHidden, setModalVisible, showModal]);
```
I've added a toggle function in useModalState.
Lets use that.

AdityaKhatri commented on src/components/general/AddStakeholderModal/SearchStakeholder/Stakeholder/index.tsx:
```diff
@@ -0,0 +1,63 @@
+        >
+            <ElementFragments
+                icons={(
+                    <Avatar
+                        className={styles.icon}
+                        src={value.logoUrl}
```
Lets provide name of organization for cases when logo is not available.

AdityaKhatri commented on src/components/general/AddStakeholderModal/SearchStakeholder/index.tsx:
```diff
@@ -0,0 +1,81 @@
+}
+
+const stakeholderKeySelector = (d: Organization) => d.id.toString();
+function SearchStakeholder(props: Props) {
```
Keep Props and function together

AdityaKhatri commented on src/newViews/ProjectEdit/ProjectDetailsForm/StakeholderList/index.tsx:
```diff
@@ -0,0 +1,86 @@
+    } = props;
+
+    if (!data || data.length === 0) {
+        return (<></>);
```
Can't we return null?

AdityaKhatri commented on src/components/general/AddStakeholderModal/StakeholderList/index.tsx:
```diff
@@ -0,0 +1,110 @@
+    } = props;
+
+    const handleDragEnter = useCallback(() => {
+    }, []);
```
This might be unnecessary.

AdityaKhatri commented on src/components/general/AddStakeholderModal/StakeholderList/index.tsx:
```diff
@@ -0,0 +1,110 @@
+import React, { useCallback } from 'react';
+import { _cs } from '@togglecorp/fujs';
+import useDropHandler from '#hooks/useDropHandler';
```
Lets use DropContainer from DEEP UI if possible.

AdityaKhatri commented on src/components/general/AddStakeholderModal/index.tsx:
```diff
@@ -0,0 +1,191 @@
+        >
+            <SearchStakeholder className={styles.left} />
+            <div className={styles.right}>
+                <StakeholderList
+                    onChange={handleChange}
+                    name="lead_organization"
+                    value={value.lead_organization}
+                    label={_ts('project.detail.stakeholders', 'leadOrganization')}
+                />
+                <StakeholderList
+                    onChange={handleChange}
+                    name="international_partner"
+                    value={value.international_partner}
+                    label={_ts('project.detail.stakeholders', 'internationalPartner')}
+                />
+                <StakeholderList
+                    onChange={handleChange}
+                    name="national_partner"
+                    value={value.national_partner}
+                    label={_ts('project.detail.stakeholders', 'nationalPartner')}
+                />
+                <StakeholderList
+                    onChange={handleChange}
+                    name="donor"
+                    value={value.donor}
+                    label={_ts('project.detail.stakeholders', 'donor')}
+                />
+                <StakeholderList
+                    onChange={handleChange}
+                    name="government"
+                    value={value.government}
+                    label={_ts('project.detail.stakeholders', 'government')}
```
This should be iterated over a list fetched from the server as organization types can be created/modified from the server.

samshara replied:
@AdityaKhatri do we have an API for this?

AdityaKhatri replied:
@samshara , there should be, could you go through the API docs or follow up with @thenav56 through slack?

thenav56 replied:
@AdityaKhatri You meant this https://api.thedeep.io/api/v1/organization-types/?

samshara replied:
@AdityaKhatri, @thenav56 I don't think the above API is what we need.

AdityaKhatri replied:
Oh okay. My apologies, I mixed them up.

thenav56 replied:
Is this resolved? or need any server-related help?

AdityaKhatri replied:
Yes this is resolved.

frozenhelium commented on src/components/general/AddStakeholderModal/SearchStakeholder/Stakeholder/index.tsx:
```diff
@@ -0,0 +1,64 @@
+        <DraggableContent
+            name="stakeholder"
+            value={dragValue}
+            className={styles.item}
```
Lets match the `className` with the name of component
```suggestion
            className={styles.stakeholder}
```

frozenhelium commented on src/components/general/AddStakeholderModal/SearchStakeholder/Stakeholder/index.tsx:
```diff
@@ -0,0 +1,64 @@
+                    />
+                )}
+            >
+                <div className={styles.text}>
```
Let's use `childrenContainerClassName={styles.text}` in `ElementsFragment` instead of adding one more `div` element

frozenhelium commented on src/components/general/AddStakeholderModal/SearchStakeholder/Stakeholder/styles.scss:
```diff
@@ -0,0 +1,45 @@
+        }
+
+        .abbr {
+            flex-grow: 1;
```
The parent container doesn't have `display: flex` so this probably won't work

frozenhelium commented on src/components/general/AddStakeholderModal/SearchStakeholder/index.tsx:
```diff
@@ -0,0 +1,82 @@
+    [searchText, stakeholders?.count]);
+
+    return (
+        <div className={_cs(className, styles.search)}>
```
Lets use appropriate `className`

frozenhelium commented on src/components/general/AddStakeholderModal/StakeholderList/index.tsx:
```diff
@@ -0,0 +1,107 @@
+    return (
+        <div
+            className={_cs(
+                styles.widget,
```
Use proper `className`

frozenhelium commented on src/components/general/AddStakeholderModal/StakeholderList/styles.scss:
```diff
@@ -0,0 +1,55 @@
+    flex-basis: calc(calc(100% / 3) - calc(var(--dui-spacing-small) * 2));
+    flex-direction: column;
+    flex-grow: 0;
+    margin: var(--dui-spacing-small);
```
We should not add `margin` from here, that job belongs to its parent

frozenhelium commented on src/components/general/AddStakeholderModal/StakeholderList/styles.scss:
```diff
@@ -0,0 +1,55 @@
+    position: relative;
+    flex-basis: calc(calc(100% / 3) - calc(var(--dui-spacing-small) * 2));
+    flex-direction: column;
+    flex-grow: 0;
```
We should not add `flex-grow` / `flex-shrink` from here, that job belongs to its parent

frozenhelium commented on src/components/general/AddStakeholderModal/StakeholderList/styles.scss:
```diff
@@ -0,0 +1,55 @@
+    flex-grow: 0;
+    margin: var(--dui-spacing-small);
+    border: var(--dui-width-separator-thin) solid var(--dui-color-separator);
+    border-radius: 5px;
```
Use appropriate variable for the `border-radius`

frozenhelium commented on src/components/general/AddStakeholderModal/index.tsx:
```diff
@@ -0,0 +1,191 @@
+
+    return (
+        <Modal
+            className={styles.modal}
```
```suggestion
            className={styles.addStakeholderModal}
```

frozenhelium commented on src/components/general/AddStakeholderModal/styles.scss:
```diff
@@ -0,0 +1,31 @@
+    height: 90vh;
+
+    .heading {
+        color: var(--dui-color-brand);
```
This should not be styled from here
CC: @AdityaKhatri

samshara replied:
@AdityaKhatri any idea on what needs to be done here?

AdityaKhatri replied:
Yes. @samshara . Lets not style it and let it be its normal form.
We can style it from Modal later on.

frozenhelium commented on src/newViews/ProjectEdit/ProjectDetailsForm/StakeholderList/styles.scss:
```diff
@@ -0,0 +1,39 @@
+                padding: var(--dui-spacing-extra-small) var(--dui-spacing-small);
+
+                .icon {
+                    width: 1.5rem;
+                    height: 1.5rem;
```
We should use already existing / new variable for `rem`
```suggestion
                    width: 1.5em;
                    height: 1.5em;
```

samshara committed:
Fix styling issues in AddStakeholderModal

samshara committed:
Separate out options and value for AddStakeholderModal

** Previously information of stakeholder(title, logoUrl) was directly
passed through value prop. Now only id's of stakeholders are passed as value.
The information about selected  stakeholder are mapped through values prop.

** Use DropContainer instead of useDropHandler

AdityaKhatri committed:
Improve styling of stakeholder modal
